File: subarrayproductlessthank.js
//Objective is to find the number of subarrays whose product is less than k


let left = 0
let right = 0
let product = 1
let result = 0
    
while (right < nums.length) {
    //Multiply the product unconditionally
    product *= nums[right]
        
    //If the product went over, then move the left pointer until it isn't
    while (product >= k) {
        product = product / nums[left++]
    }
        
    //Update the result every subarray
    result += right - left + 1
        
    right++
}
    
//If the result ends up being negative, return 0 instead
return result < 0 ? 0 : result

File: subarrayproductlessthank.java
//Java Solution

class Solution {
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        int left = 0;
        int right = 0;
        int product = 1;
        int ans = 0;
        
        while (right < nums.length) {
            product *= nums[right];
            
            while (left < nums.length && product >= k) {
                product /= nums[left++];
            }
            
            ans += right - left + 1;
            right++;
        }
        
        return ans > 0 ? ans : 0;
    }
}